run_foundry.py

```diff
@@ -0,0 +1,478 @@
+import argparse
+import glob
+from os.path import join
+from subprocess import CompletedProcess
+from typing import Mapping
+
+from hypothesis import given, settings, Verbosity
+from hypothesis.strategies import integers, tuples
+from pyk.prelude.utils import token
+from pyk.prelude.collections import list_of, map_of
+from pyk.kast.inner import KSort, KVariable
+from pyk.kast.kast import kast_term
+from pyk.kore.syntax import Pattern
+from pyk.ktool.kprint import _kast, KAstInput, KAstOutput
+from pyk.ktool.krun import _krun, KRunOutput
+from pyk.cterm import CTerm, build_claim
+from pyk.utils import ensure_dir_path
+from pyk.prelude.kint import leInt
+from pyk.prelude.ml import mlEqualsTrue
+
+from run_elrond_tests import *
+
+INPUT_FILE_NAME = 'foundry.json'
+
+def load_input_json(test_dir):
+    try:
+        with open(join(test_dir, INPUT_FILE_NAME), 'r') as f:
+            return json.load(f)
+    except FileNotFoundError as e:
+        raise FileNotFoundError('"{INPUT_FILE_NAME}" not found in "{test_dir}"')
+
+
+def load_wasm(filename):
+    with open(filename, 'rb') as f:
+        return wasm2kast.wasm2kast(f, filename)
+
+
+def find_test_wasm_path(test_dir):
+    test_wasm_path = glob.glob(test_dir + '/output/*.wasm')
+    # TODO this loads the first wasm file in the directory. what if there are multiple wasm files?
+    if test_wasm_path:
+        return test_wasm_path[0]
+    else:
+        raise ValueError(f'WASM file not found: {test_dir}/output/?.wasm')
+
+
+def load_contract_wasms(contract_wasm_paths):
+    contract_wasm_modules = {
+        bytes(f, 'ascii'): load_wasm(f) for f in contract_wasm_paths
+    }
+    return contract_wasm_modules
+
+
+def deploy_test(krun, test_wasm, contract_wasms):
+    """
+    1. create a main account: 'k'
+    2. reserve a new address for the test contract: owner = 'k', contract address = 'k-test'
+    3. deploy the test contract from account 'k': 'k-test'
+    """
+    # create the root account
+    k_addr = mandos_argument_to_kbytes('address:k')
+    init_main_acct = KApply(
+        'setAccount',
+        [
+            k_addr,
+            token(1),
+            token(100000),
+            KApply('.Code', []),
+            token(b''),
+            KMapBytesToBytes([]),
+        ],
+    )
+
+    # the test contract's address will be 'k-test'
+    k_test_addr = mandos_argument_to_kbytes('sc:k-test')
+    new_address = KApply('newAddress', [k_addr, token(1), k_test_addr])
+
+    # deploy the test contract
+    arguments = ListBytes(wrapBytes(token(k)) for k in contract_wasms)
```

Comment:
I'm not sure I understand this - Why do we send the WASM code for some contracts as arguments to the `init` function? Is it because we expect a test's init function to get other contracts as arguments, and nothing else?

Reply:
We pass contract path(s) to the test contract, not the whole contract. Paths serve as identifiers for contracts.
The idea is that the VM keeps a map of contract IDs (i.e. paths) to contract codes, and we pass those IDs to the test contract so that it can send deployment commands to the VM, "deploy contract A, deploy contract B" etc.

Reply:
Ok, thanks.